Kenesbek Proj

serzhan-kenesbek opened the pull request:
The game is playable in the Terminal. 

There are currently no Tests yet, as I wanted to first get feedback on the implementation before creating them. 

In Explanations.txt file I documented my thinking process when creating most features. Will later turn it into a proper documentation.

serzhan-kenesbek committed:
Add most of game logic and features

serzhan-kenesbek committed:
Changed Explanations.txt a bit

kajigor commented:
There's something going wrong: it sometimes hangs, for example: 

![Screenshot 2024-11-27 at 12 20 38](https://github.com/user-attachments/assets/d44712ba-1f6c-4734-8f4c-ea92c4d3b720)

kajigor commented on minesweeper/src/Grid/Generation.hs:
```diff
@@ -0,0 +1,73 @@
+-- Count mines around a specific cell
+countAdjacentMines :: [Cell] -> Int -> Int -> Int -> Int
+countAdjacentMines flatGrid rows cols index =
+  length [ () | neighbor <- getNeighbors index rows cols, isMine (flatGrid !! neighbor) ]
```
https://hackage.haskell.org/package/ghc-9.10.1/docs/GHC-Utils-Misc.html#v:count

kajigor commented on minesweeper/src/Grid/Core.hs:
```diff
@@ -0,0 +1,33 @@
+renderGrid grid = do
+    putStrLn $ "    " ++ unwords (map (pad . show) [0 .. cols - 1]) -- Column indices
+    putStrLn $ "  " ++ replicate horizontalLineLength '-'           -- Top border
+    mapM_ renderRow (zip ([0 ..] :: [Int]) grid) -- Had a warning because mapM_ can't infer the type of the list being passed to it. That's why the list passed to zip is explicitly Int
```
You could also specify the type of `renderRow`.

kajigor commented on minesweeper/src/Grid/Core.hs:
```diff
@@ -0,0 +1,33 @@
+
+    -- Render a single cell of the grid
+    renderCell cell
+        | isRevealed cell && isMine cell  = "*   " -- Revealed mine
+        | isRevealed cell                 = show (adjacentMines cell) ++ "   "
```
It doesn't matter in a grand scheme of things, but I'd reorder the `isMine` and `isRevealed` check to avoid doing the latter twice.

kajigor commented on Explanations.txt:
```diff
@@ -0,0 +1,92 @@
+  return $ take n (nub indices)
+
+wanted to do this, but then if indices are duplicated and removed due to nub there would be less than n mines
+Instead used Sets since they already handle the duplicates.
```
I bet there's a way to repeat monadic action infinitely many time in a lazy fashion and then take enough elements from the prefix to have n unique numbers.

kajigor commented on minesweeper/src/Game.hs:
```diff
@@ -0,0 +1,119 @@
+        Nothing -> return $ Left "Invalid input format. Use 'rows,cols,mines'."
+
+-- Parse the input string for game parameters
+parseGameParameters :: String -> Maybe (Int, Int, Int)
+parseGameParameters input =
+    case map readMaybe (splitOn "," input) of
+        [Just rows, Just cols, Just mines] -> Just (rows, cols, mines)
+        _ -> Nothing
```
I'd also skipped whitespace between inputs.

kajigor commented on minesweeper/src/Game.hs:
```diff
@@ -0,0 +1,119 @@
+validateParameters rows cols mines
+    | rows <= 0 || cols <= 0 = Left "Grid dimensions must be greater than 0."
+    | mines <= 0 = Left "Number of mines must be greater than 0."
+    | mines >= rows * cols = Left "Number of mines must be less than the total number of cells."
```
There's nothing wrong with placing a mine in every cell 🥲

kajigor commented on Explanations.txt:
```diff
@@ -0,0 +1,92 @@
+First playable version ready. 
+
+After playing for a bit found a lot of issues.
+One of them was that I forgot to take into account that the first move that the player makes should never turn out to be a mine.
```
Well... The version of a minesweeper game I had in my childhood allowed the user to blow up at the first move.

kajigor commented on minesweeper/src/Game.hs:
```diff
@@ -0,0 +1,119 @@
+                    putStrLn "Invalid move: Cell is out of bounds. Try again!"
+                    gameLoop grid rows cols
+                else do
+                    let updatedGrid = if isFlag then flagCell row col grid else revealCell row col grid
+                    if isFlag
+                        then gameLoop updatedGrid rows cols
+                        else checkGameState updatedGrid row col rows cols
```
Excessive checks

kajigor commented on minesweeper/src/Game.hs:
```diff
@@ -0,0 +1,119 @@
+    putStrLn "\nEnter your move (e.g., 'row,column' to reveal or 'F row,column' to flag):"
+    input <- getLine
+
+    case parseInput input of
```
I'd introduce a function

kajigor commented on minesweeper/src/Grid/Core.hs:
```diff
@@ -0,0 +1,33 @@
+import Cell (Cell(..))
+
+-- Define the Grid type as a 2D list of Cell
+type Grid = [[Cell]]
```
Using lists may be a tad slow. Can you explore which dimensions cause a noticeable lag?

serzhan-kenesbek committed:
Improve based on feedback, add tests and documentation

serzhan-kenesbek commented:
Made changes as requested, fixed and understood the first bug, added testing (both Unit and Property-Based), and added final documentation in report.pdf

serzhan-kenesbek committed:
Added documentation

kajigor commented:
Blows up: 

```
Ekaterina.Verbitskaya@NVC00653 minesweeper % stack run
Enter grid dimensions and number of mines (e.g., rows,cols,mines):
3,1,2
      0
  --------
  0 | .   |
  --------
  1 | .   |
  --------
  2 | .   |
  --------

Enter your first move (row,column):
1,0
minesweeper: Too many mines: Cannot place 2 mines in a grid with only 0 valid cells.
CallStack (from HasCallStack):
  error, called at src/Grid/Generation.hs:64:31 in minesweeper-0.1.0.0-2hDQarTrxVG3Aaa6xdcYbH:Grid.Generation
```

serzhan-kenesbek commented:
It is intended behavior. If there are too many mines it throws an error

kajigor commented:
Throwing exceptions into a user must never be done. If you, for some reason, think that 2 mines don't fit into 3 cells, warn the user, and do it when creating the grid, not after the first move.

serzhan-kenesbek commented:
Yeah makes sense. My bad. It's just that having a 3x1 grid with two mines would mean that you win with your first move. This way I was also trying to prevent the bug that you noticed when I submitted previously

serzhan-kenesbek committed:
Fix

serzhan-kenesbek commented:
Fixed it

kajigor commented:
Ok. Good job 👍
